frontend/assessments/3/src/Bookingform.tsx

```
@@ -0,0 +1,131 @@
+import { RoomType, AddOn } from './redux/Roomtypes';
+import { fetchRoomTypesAndAddOns } from './redux/FetchRoomtypes';
+import { RootState } from './redux/Roomtypes';
+import "./Bookingform.scss";
+
+const BookingForm: React.FC = () => {
+ const dispatch = useDispatch();
+ const state = useSelector((state: RootState) => state); 
+ const { roomTypes, selectedRoomType, startDate, endDate, selectedAddOns } = state.booking;
+ const [isFormValid, setIsFormValid] = useState(false);
+ useEffect(() => {
+    const fetchData = async () => {
+      const data = await fetchRoomTypesAndAddOns();
+      dispatch(setRoomTypes(data.roomTypes));
+    };
+
+    fetchData();
+ }, [dispatch]);
+
+ useEffect(() => {
+    const isValid = selectedRoomType !== null && startDate !== null && endDate !== null;
+    setIsFormValid(isValid);
+}, [selectedRoomType, startDate, endDate]);
+
+ const handleRoomTypeChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    const selectedRoomTypeId = e.target.value;
+    const selectedRoomType = roomTypes.find((type:RoomType) => type.id.toString() === selectedRoomTypeId);
+    if (selectedRoomType) {
+      dispatch(setSelectedRoomType(selectedRoomType));
+    }
+ };
+
+ const handleStartDateChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    dispatch(setStartDate(e.target.value));
+ };
+
+ const handleEndDateChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    dispatch(setEndDate(e.target.value));
+ };
+
+
+ const handleAddOnsChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    const selectedAddOnNames = Array.from(e.target.selectedOptions, option => option.value); 
+    const selectedAddOns: AddOn[] = selectedRoomType?.addOns.filter((addOn:AddOn) => selectedAddOnNames.includes(addOn.name)) || [];  
+    dispatch(setSelectedAddOns(selectedAddOns));
+};
+
+const calculateTotalCost = (): number => {
+    if (!selectedRoomType || !startDate || !endDate) {
+       return 0; 
```

not 0 should return an error messge.